### PR TITLE
WIP refactor container and queuing system

### PR DESCRIPTION
**Pull request by alextran1502**
Refactor the server-side of the application into monorepo architecture with microservices handled in a queue system, separate from the main server to ensure high I/O is handled smoothly.

The queue system is inspired by @zackpollard with the diagram below

queue-diagram

This PR would be treated as breaking changes to the system (although no user data will be lost). Users will need to update the docker-compose file to restore all of the functionality of the app

### CODE REVIEW

File: server/apps/immich/src/api-v1/asset/asset.controller.ts [hunk 1]
```diff
@@ -61,13 +66,24 @@ export class AssetController {
         const savedAsset = await this.assetService.createUserAsset(authUser, assetInfo, file.path, file.mimetype);
```
Comment (alextran1502) [inline server/apps/immich/src/api-v1/asset/asset.controller.ts:67]: This function now only handles saving the new asset info into the data (thumbnail path if available) and then putting a job into the `asset-uploaded` queue. The first stage in the diagram.

immich-queue-layour copy
m

File: server/apps/microservices/src/processors/asset-uploaded.processor.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+   * Post processing uploaded asset to perform the following function if missing
+   * 1. Generate JPEG Thumbnail
+   * 2. Generate Webp Thumbnail <-> if JPEG thumbnail exist
+   * 3. EXIF extractor
+   * 4. Reverse Geocoding
+   *
+   * @param job asset-uploaded
+   */
+  @Process('asset-uploaded')
+  async processUploadedVideo(job: Job) {
```
Comment (alextran1502) [inline server/apps/microservices/src/processors/asset-uploaded.processor.ts:34]: This is the starting point of the microservice that handles post-upload-related tasks. It will then put the additional jobs into the corresponding queues.

* Generate JPEG Thumbnail
* Generate Webp Thumbnail <-> if JPEG thumbnail exists
* EXIF extractor
* Reverse Geocoding
* Video Encoding
   
This is the second stage in the diagram.

immich-queue-layour copy 2

File: server/apps/microservices/src/processors/video-transcode.processor.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import { APP_UPLOAD_LOCATION } from '../../../immich/src/constants/upload_location.constant';
+
+@Processor('video-conversion-queue')
+export class VideoTranscodeProcessor {
+  constructor(
+    @InjectRepository(AssetEntity)
+    private assetRepository: Repository<AssetEntity>,
+  ) {}
+
+  @Process({ name: 'mp4-conversion', concurrency: 1 })
```
Comment (alextran1502) [inline server/apps/microservices/src/processors/video-transcode.processor.ts:18]: The `concurrency: 1` option limits one worker in the pool to handle this process, any additional videos that are added to the queue will be in the waiting queue until the worker is idle.

File: server/apps/microservices/src/processors/video-transcode.processor.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      const savedEncodedPath = encodedVideoPath + '/' + asset.id + '.mp4';
+
+      if (asset.encodedVideoPath == '' || !asset.encodedVideoPath) {
+        // Put the processing into its own async function to prevent the job exist right away
+        await this.runFFMPEGPipeLine(asset, savedEncodedPath);
+      }
+    }
+  }
+
+  async runFFMPEGPipeLine(asset: AssetEntity, savedEncodedPath: string): Promise<void> {
```
Comment (alextran1502) [inline server/apps/microservices/src/processors/video-transcode.processor.ts:39]: This job at first didn't conform to the concurrency setting due to the callback functions of the FFmpeg command. I have to refactor to a separate async function so that the `process` function return after the job is either success or failed

**Comment by alextran1502** [comment]
Hey, @zackpollard This PR is completed in terms of refactoring to microservices and the queue system. There seems to be a lot of files changed due to the refactoring of the project structure into monorepo to share codes between the microservices and the core server.

It would be easier to see the changes if you pull it to your local.

I've added some comments on places that matter the most in this PR. If you have time, can you take a look and leave some feedback?

Thanks,
Alex

**Comment by zackpollard** [comment]
Looks very solid to me, I haven't had time to review all the code, but I took a look at the main parts, adding to the original asset uploaded queue, the task that creates all the other messages off the back of that and a couple of the individual jobs and it looks pretty solid.

File: server/apps/microservices/src/microservices.module.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    BullModule.forRootAsync({
+      useFactory: async () => ({
+        redis: {
+          host: process.env.REDIS_HOSTNAME || 'immich_redis',
+          port: 6379,
+        },
+      }),
+    }),
+    BullModule.registerQueue({
+      name: 'thumbnail-generator-queue',
```
Comment (zackpollard) [inline server/apps/microservices/src/microservices.module.ts:28]: There's two definitions for this queue, is that intended?

File: server/apps/microservices/src/microservices.module.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    BullModule.registerQueue({
+      name: 'metadata-extraction-queue',
+      defaultJobOptions: {
+        attempts: 3,
+        removeOnComplete: true,
+        removeOnFail: false,
+      },
+    }),
+    BullModule.registerQueue({
+      name: 'video-conversion-queue',
```
Comment (zackpollard) [inline server/apps/microservices/src/microservices.module.ts:52]: There's two definitions for this queue, is that intended?

File: server/apps/microservices/src/processors/asset-uploaded.processor.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+      asset,
+      fileName,
+      fileSize,
+      hasThumbnail,
+    }: { asset: AssetEntity; fileName: string; fileSize: number; hasThumbnail: boolean } = job.data;
+
+    if (hasThumbnail) {
+      // The jobs below depends on the existence of jpeg thumbnail
+      await this.thumbnailGeneratorQueue.add('generate-webp-thumbnail', { asset }, { jobId: randomUUID() });
+      await this.metadataExtractionQueue.add('tag-image', { asset }, { jobId: randomUUID() });
```
Comment (zackpollard) [inline server/apps/microservices/src/processors/asset-uploaded.processor.ts:45]: tag-image and detect-object should happen regardless of whether the thumbnail exists

File: server/apps/microservices/src/processors/thumbnail.processor.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  ) {}
+
+  @Process('generate-jpeg-thumbnail')
+  async generateJPEGThumbnail(job: Job) {
+    const { asset }: { asset: AssetEntity } = job.data;
+
+    console.log(asset);
+  }
+
+  @Process('generate-webp-thumbnail')
```
Comment (zackpollard) [inline server/apps/microservices/src/processors/thumbnail.processor.ts:22]: What is the default concurrency for all these jobs? Generating the WebP thumbnails could still be quite intensive so probably don't want the background tasks going crazy and taking CPU, at least before it can be configurable for the user

File: server/Dockerfile [hunk 9]
```diff
@@ -8,7 +8,7 @@ COPY package.json package-lock.json ./
 
 RUN apk add --update-cache build-base python3 libheif vips-dev vips ffmpeg
 
-RUN npm install
+RUN npm install --legacy-peer-deps
```
[the remark this answers is not in the document]
Comment (alextran1502) [reply server/Dockerfile:11]: This is the build issue from Github Action that required this to be resolved.

[the remark this answers is not in the document]
Comment (alextran1502) [reply server/Dockerfile:11]: Particular this one https://github.com/alextran1502/immich/runs/6766067065?check_suite_focus=true

File: server/apps/immich/src/api-v1/asset/asset.controller.ts [hunk 10]
```diff
@@ -61,13 +66,24 @@ export class AssetController {
...
 
         if (uploadFiles.thumbnailData != null && savedAsset) {
```
[the remark this answers is not in the document]
Comment (alextran1502) [reply server/apps/immich/src/api-v1/asset/asset.controller.ts:68]: This makes sense, it should only check for the thumbnail is there or not, if there the process of saving an asset has a problem, it would throw an error in the `assetService`, and so this check won't be necessary.

File: server/apps/immich/src/api-v1/asset/asset.controller.ts [hunk 11]
```diff
@@ -61,13 +66,24 @@ export class AssetController {
...
+            uploadFiles.thumbnailData[0].path,
+          );
+
+          await this.assetUploadedQueue.add(
+            'asset-uploaded',
+            { asset: assetWithThumbnail, fileName: file.originalname, fileSize: file.size, hasThumbnail: true },
+            { jobId: savedAsset.id },
+          );
+        } else {
+          await this.assetUploadedQueue.add(
```
[the remark this answers is not in the document]
Comment (alextran1502) [reply server/apps/immich/src/api-v1/asset/asset.controller.ts:80]: This is nice but I think it applies to this case since the asset payload property of `asset` received two different payloads with different content. The content will be used later in the queue. If we try to keep the if statement simple, the refactor will be like this 

```typescript
        let assetWithThumbnail: AssetEntity;

        const savedAsset = await this.assetService.createUserAsset(authUser, assetInfo, file.path, file.mimetype);

        if (uploadFiles.thumbnailData != null) {
          assetWithThumbnail = await this.assetService.updateThumbnailInfo(
            savedAsset,
            uploadFiles.thumbnailData[0].path,
          );
        }

        await this.assetUploadedQueue.add(
          'asset-uploaded',
          {
            asset: uploadFiles.thumbnailData != null ? assetWithThumbnail : savedAsset,
            fileName: file.originalname,
            fileSize: file.size,
            hasThumbnail: uploadFiles.thumbnailData != null,
          },
          { jobId: savedAsset.id },
        );
```

There is less but also less verbose I think.

File: server/apps/immich/src/modules/schedule-tasks/schedule-tasks.module.ts [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+@Module({
+  imports: [
+    TypeOrmModule.forFeature([AssetEntity]),
+    BullModule.registerQueue({
+      name: 'thumbnail-generator-queue',
+      defaultJobOptions: {
+        attempts: 3,
+        removeOnComplete: true,
+        removeOnFail: false,
```
[the remark this answers is not in the document]
Comment (alextran1502) [reply server/apps/immich/src/modules/schedule-tasks/schedule-tasks.module.ts:15]: It will sit in the failed job section. So the mechanism of retrying the job is on us, whether to listening to the failed event and put it back to the queue of have a cronjob to do that (we are using cronjob for two most important jobs - the Wepb generator and video encoding, will be thumbnail generator later on as well)

File: server/apps/microservices/src/processors/asset-uploaded.processor.ts (hunk 7 above)
Comment (alextran1502) [reply server/apps/microservices/src/processors/asset-uploaded.processor.ts:45]: The function that decodes the image into tensors of TensorFlow expects these formats `BMP, GIF, JPEG, and PNG`. So if you upload a raw formats (HEIC, HEIF, or DNG), this function won't be able to run

File: server/apps/microservices/src/processors/thumbnail.processor.ts (hunk 8 above)
Comment (alextran1502) [reply server/apps/microservices/src/processors/thumbnail.processor.ts:22]: The concurrency for these jobs, if not specified, is set to the maximum allowable workers, which is the available Redis Client, I think it would make sense to set the level as 2 or so in this case

https://github.com/OptimalBits/bull/blob/develop/REFERENCE.md#queuegetworkers

File: server/apps/microservices/src/microservices.module.ts (hunk 5 above)
Comment (alextran1502) [reply server/apps/microservices/src/microservices.module.ts:28]: There is no duplication on the same module. But you have to define the Queue on the module that uses it, i.e `schedule-tasks.module.ts`, otherwise you get this error 

```bash
immich-server_1            | [Nest] 158  - 06/09/2022, 1:18:43 AM   ERROR [ExceptionHandler] Nest can't resolve dependencies of the ScheduleTasksService (AssetEntityRepository, ?, BullQueue_video-conversion-queue). Please make sure that the argument BullQueue_thumbnail-generator-queue at index [1] is available in the ScheduleTasksModule context.
immich-server_1            | 
immich-server_1            | Potential solutions:
immich-server_1            | - If BullQueue_thumbnail-generator-queue is a provider, is it part of the current ScheduleTasksModule?
immich-server_1            | - If BullQueue_thumbnail-generator-queue is exported from a separate @Module, is that module imported within ScheduleTasksModule?
immich-server_1            |   @Module({
immich-server_1            |     imports: [ /* the Module containing BullQueue_thumbnail-generator-queue */ ]
immich-server_1            |   })
immich-server_1            | 
immich-server_1            | Error: Nest can't resolve dependencies of the ScheduleTasksService (AssetEntityRepository, ?, BullQueue_video-conversion-queue). Please make sure that the argument BullQueue_thumbnail-generator-queue at index [1] is available in the ScheduleTasksModule context.
immich-server_1            | 
immich-server_1            | Potential solutions:
immich-server_1            | - If BullQueue_thumbnail-generator-queue is a provider, is it part of the current ScheduleTasksModule?
immich-server_1            | - If BullQueue_thumbnail-generator-queue is exported from a separate @Module, is that module imported within ScheduleTasksModule?
immich-server_1            |   @Module({
immich-server_1            |     imports: [ /* the Module containing BullQueue_thumbnail-generator-queue */ ]
immich-server_1            |   })
immich-server_1            | 
immich-server_1            |     at Injector.lookupComponentInParentModules (/usr/src/app/node_modules/@nestjs/core/injector/injector.js:202:19)
immich-server_1            |     at processTicksAndRejections (node:internal/process/task_queues:96:5)
immich-server_1            |     at Injector.resolveComponentInstance (/usr/src/app/node_modules/@nestjs/core/injector/injector.js:157:33)
immich-server_1            |     at resolveParam (/usr/src/app/node_modules/@nestjs/core/injector/injector.js:108:38)
immich-server_1            |     at async Promise.all (index 1)
immich-server_1            |     at Injector.resolveConstructorParams (/usr/src/app/node_modules/@nestjs/core/injector/injector.js:123:27)
immich-server_1            |     at Injector.loadInstance (/usr/src/app/node_modules/@nestjs/core/injector/injector.js:52:9)
immich-server_1            |     at Injector.loadProvider (/usr/src/app/node_modules/@nestjs/core/injector/injector.js:74:9)
immich-server_1            |     at async Promise.all (index 3)
immich-server_1            |     at InstanceLoader.createInstancesOfProviders (/usr/src/app/node_modules/@nestjs/core/injector/instance-loader.js:44:9)
```

Comment (alextran1502) [reply server/apps/microservices/src/microservices.module.ts:28]: Actually I lied... From the `microservicess.module.ts`, you can export `BullModule`. Then in `schedule-tasks.module.ts`, including the `MicroservicesModule`, then you don't have to register the queue again.

File: server/apps/microservices/src/microservices.module.ts (hunk 6 above)
Comment (alextran1502) [reply server/apps/microservices/src/microservices.module.ts:52]: Resolved above.

File: server/Dockerfile (hunk 9 above)
[the remark this answers is not in the document]
Comment (zackpollard) [reply server/Dockerfile:11]: Ah I see, with yarn you can fix this with custom resolutions, unsure what the procedure is for npm as I don't use npm for any of my projects anymore

File: server/apps/immich/src/api-v1/asset/asset.controller.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (zackpollard) [reply server/apps/immich/src/api-v1/asset/asset.controller.ts:80]: Ah I see, yes I missed the different asset usage, original way is all good then, easier to understand what's going on that way

File: server/apps/immich/src/modules/schedule-tasks/schedule-tasks.module.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (zackpollard) [reply server/apps/immich/src/modules/schedule-tasks/schedule-tasks.module.ts:15]: Sounds good, just don't want failed jobs to get queued up in the background and fill the queues up causing memory issues

File: server/apps/microservices/src/microservices.module.ts (hunk 5 above)
Comment (zackpollard) [reply server/apps/microservices/src/microservices.module.ts:28]: Fair enough, it seems weird to me that you have to redefine it along with the config too, as which config does it decide to take?

File: server/apps/microservices/src/processors/thumbnail.processor.ts (hunk 8 above)
Comment (zackpollard) [reply server/apps/microservices/src/processors/thumbnail.processor.ts:22]: Yea I think setting some sensible limit would be good just so it doesn't eat CPU. Longer term I think these options arre great candidates for the admin settings page as some kind of advanced options.

File: server/apps/microservices/src/microservices.module.ts (hunk 5 above)
Comment (alextran1502) [reply server/apps/microservices/src/microservices.module.ts:28]: This is the part I am unsure of, so I redefine the queue to have the same setting just in case.

File: server/apps/microservices/src/processors/asset-uploaded.processor.ts (hunk 7 above)
Comment (alextran1502) [reply server/apps/microservices/src/processors/asset-uploaded.processor.ts:45]: Yes, but it is also the trade-off between CPU (or GPU) usage. Below is a good read/discussion about this topic

https://stackoverflow.com/questions/45149023/tensorflow-for-image-recognition-size-of-images

Comment (zackpollard) [reply server/apps/microservices/src/processors/asset-uploaded.processor.ts:45]: That's talking specifically about training the model, right? Presumably you're using a pre-trained model and just pushing images into it for analysis and tagging, I would imagine for that task using the full resolution image is preferable, especially if that's what the data was trained on.

Comment (alextran1502) [reply server/apps/microservices/src/processors/asset-uploaded.processor.ts:45]: Correct, the full resolution would take more CPU to run through all the pixels, so it is the trade-off between accuracy and CPU usage. I guess we can also throttle this queue with the concurrency level as well. I guess we can have this feature in the future as a setting in the admin panel as well since other people might have their own model that takes a specific resolution. So the ML pipeline can generate the "thumbnail" image with the specified resolution from the user.

Comment (zackpollard) [reply server/apps/microservices/src/processors/asset-uploaded.processor.ts:45]: Yea for sure, would be interested to see how much more CPU time it takes to run on the full resolution image, with my understanding of how the models tend to work I wouldn't expect it to actually be that much more intensive for analysis, just for training. Can take a look at that later though, was just interesting to see how this works.
